README
 Usage Notes: 

 - To let an application use FCOM

    o #include <fcom_api.h>
	o link with -lfcom
    o link with -ludpCommBSD (for FCOM over the regular TCP/IP stack)
    o link with -ludpComm    (for FCOM over the dedicated stack -- RTEMS ONLY)

 - Programming (API) documentation is available in 'doc/fcom_icd.pdf'.

 - OS setup information 
   --------------------

   On machines with multiple NICs a routing table entry should be added
   so that multicast traffic goes over the correct interface.

      RTEMS: rtconf(1, <fcom-mcast-prefix>, <ip-of-interface>, <mcast-prefix-netmask>)

      LINUX: route add -net <fcom-mcast-prefix> netmask <mcast-prefix-netmask> dev <nic-name>
          


ChangeLog
2011-12-12 (T.S.)

	* src/xdr_swpP.h: Apparently, __builtin_bswap32() appeared with
    gcc-4.3. Modified version checking macros accordingly.
	Also: use __builtin_bswap32() on all LE architectures (not just i386).

2010-04-26 (T.S.)

	* icd/Makefile, icd/fcom_icd.cksum: Use checksums instead
	of file date to check if we need to remake the PDF file.
	Otherwise, the order of a CVS checkout may trigger a rebuild...

2010-04-26 (T.S.)

	* ChangeLog: fixed bad dates.
	* RELEASE fcom-R1-3-0 cut.

2010-04-22 (T.S.)

	* src/fcget.c:
	 o New '-b <buffer_count>' option to set/increase # of buffers
	   (in case the user passes many IDs).
	 o New '-s' option to dump statistics before exiting.
	 o BUGFIX: If this is an asynchronous 'get' then timeout arg
	           to fcomGetBlob() must be zero.
	
2010-04-22 (T.S.)

	* src/fc_recvr.c: Added some statistics for sets. More paranoia
	checking of arguments (let fcomFreeBlobSet accept and ignore
	NULL set pointer; let fcomAllocBlobSet() NULL the returned 
	set handle on failure.

2010-04-22 (T.S.)

	* src/Makefile, src/fc_recvr.c: 'set' support is now a compile-time
	option. Define SUPPORT_SETS to enable it (added -DSUPPORT_SETS to
	Makefile).

2010-04-21 (T.S.)

	* icd/fcom_icd.tex, icd/fcom_api.h, icd/fcom_icd.pdf: Updated
   documentation ( adding new 'set' feature and fcomDumpBlob() ).

2010-04-21 (T.S.)

	* src/fcget.c: Accept multiple IDs on command line and use a blob set
	if the user gave more than one.
	Replaced call to fcomDumpIDStats() by fcomDumpBlob().

2010-04-21 (T.S.)

	* src/fc_recv.c:
	 o implemented support for 'blob sets' (THIS FEATURE REQUIRES PTHREADS)
	   which allows a single thread to block for arrival of a set of blobs.

	 o BUGFIX: fixed a few places where a negative errno status was passed
	   to FCOM_ERR_SYS(errno). This macro expects a positive error status.

	 o added new (internal) routine to compute condvar timeouts. This now
	   uses clock_gettime() if available (testing _POSIX_TIMERS macro) and
	   falls back on gettimeofday() otherwise.

	 o added fcomDumpBlob() (factored most code out of fcomDumpIDStats() and
	   made the latter a wrapper around fcomDumpBlob()).

2010-04-21 (T.S.)

	* src/xdr_swpP.h: recognize __ppc__ (OSX). Added check and #error
	if neither __LITTLE_ENDIAN__ nor __BIG_ENDIAN__ is defined and
	architecture is not explicitly recognized.

2010-04-21 (T.S.)

	* src/fcomxtst.c, src/fcomrtst.c, src/fcometst.c, src/fcomitst.c:
	Check FCOM_MC_PREFIX environment variable and pass to fcomInit() if set.

	* src/fcomxtst.c: print message indicating that there might be syntax errors
	if scanning a line of input file fails.

2010-04-21 (T.S.)

	* src/fcomrtst.c: Include <unistd.h> for getopts on OSX.

2010-04-21 (T.S.)

	* src/fc_strerror.c: Added description string for new error code FCOM_ERR_ID_IN_USE.

2010-04-21 (T.S.)

	* src/blobio.c: Added short description of file format to header.

2010-04-21 (T.S.)

	* icd/fcom_api.h:
	 o include <inttypes.h> and define format string (FCOM_ID_FMT) for printing IDs.
	 o added new error code FCOM_ERR_ID_IN_USE.
	 o added declaration of new fcomDumpBlob() entry point (identical to
	   fcomDumpIDStats() but a given blob handle may be passed instead of an ID
	   thus avoiding a lookup).
	 o added API declaration of new 'blob set' feature which allows a single thread
	   to block for arrival of a set of blobs.

2010-04-21 (T.S.)

	* src/Makefile: made test programs PROD_HOST instead of PROD_Linux.

2010-03-19 (T.S.)

	* src/fc_init.c, src/fc_recv.c, src/fcomP.h:
	Added 'fcom_silent_mode' to suppress warnings about non-RT scheduler
	and mutex priority inheritance not being available.
	Certain (diagnostic) applications don't care.
	However, this is not a public feature for now.

2010-03-19 (T.S.)

	* src/fcget.c: Added '-i' swith / FCOM_MC_IFADDR env-var to
	define interface where we should listen for FCOM

2010-03-19 (T.S.)

	* src/fcometst.c, src/prototst.c, src/mainwrap.h: Moved
	wrapper code (convert vararg list of strings to proper array
	that can be passed to a 'main'-style routine) to 'mainwrap.h'
	to be used by multiple test programs.

2010-03-19 (T.S.)

	* src/Makefile, src/fcget.c: New host diagnostic/test
	program for reading data from FCOM.

2010-03-19 (T.S.)

	* src/fc_recv.c, icd/fcom_icd.tex, icd/fcom_api.h: Added
	'fcomDumpIDStats()' diagnostic routine and documented in
	the ICD.

2010-03-18 (T.S.)

	* src/mainwrap.h, src/prototst.c, src/fcomrtst.c: separated
	wrapper around 'main' for RTEMS into separate header that
	can be used by different utilites.

	* src/fcomrtst.c: added more command line options/features.

	* src/fcomitst.c, src/Makefile: added 'fcomitst' test program
	for interactive operation.

2010-03-17 (T.S.)

	* src/xdr_enc.c:
	BUGFIX: PPC dcbz optimization cleared source memory instead
	of destination.

2010-01-27 (T.S.)

	* src/Makefile, src/fcom_iocsh.c, src/fcomIocshSupport.dbd: 
	The latter two are new files used to implement a library
	'libfcomIocshSupport' that registers (so far)
	 o fcomInit()
	 o fcomDumpStats()
	 o fcomStrerror()
	with iocsh().

2010-01-27 (T.S.)

	* src/fc_send.c: BUGFIX: check for version number embedded in
	ID (fcomAllocGroup()) used the wrong macros (letting fcomAllocGroup()
	fail).

2010-01-13 (T.S.)

	* src/fcometst.c:
	  o fcomPingBlob() - check return status of fcomPutBlob()
	    and notify the user of errors.
	  o fcomPingBlob() - added note about the inherent race condition
	    (on slow machines).
	  o BUGFIX: call fromInit() in all cases with a nonzero number of
	    buffers (RX is always used!).
	  o Added 'sleep()' after fcomSubscribe() to give the subscription
	    process and the switch(es) some time before starting the real test.

2010-01-13 (T.S.)

	* src/prototst.c: newlib/rtems-4.10 must now #define __need_getopt_newlib
	in order to get declarations for getopt_r().

2010-01-13 (T.S.)

	* src/fc_send.c, src/fc_recv.c: added (compile-time optional) code for
	profiling/timing.

2010-01-13 (T.S.)


src/fcget.c
/*$Id: fcget.c,v 1.4 2010/04/22 02:16:31 strauman Exp $*/

/* Get FCOM Blob and dump to stdout */

#include <fcom_api.h>
#include <udpComm.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <inttypes.h>
#include <time.h>

#include <fcomP.h> /* for fcom_silent_mode which is not really public */


static void
usage(char *nm)
{
	fprintf(stderr,"Usage: %s [-ahvs] [-t <timeout_ms>] [-p <fcom_mc_prefix>] [-i <fcom_mc_IF>] [-b bufs] blob_id {blob_id}\n", nm);
	fprintf(stderr,"  Options:\n");
	fprintf(stderr,"       -h print this message\n");
	fprintf(stderr,"       -a enforce asynchronous 'get'\n");
	fprintf(stderr,"       -b configure number of buffers (if you use many IDs); default=10\n");
	fprintf(stderr,"       -t <timeout_ms> to wait for new data (or delay\n");
	fprintf(stderr,"          after subscription until attempting async get).\n");
	fprintf(stderr,"          Defaults to 1000.\n");
	fprintf(stderr,"       -p <fcom_mc_prefix>. Multicast prefix for FCOM\n");
	fprintf(stderr,"       -i <fcom_mc_IF>. IF (dot-address) on which to listen for FCOM\n");
	fprintf(stderr,"       -v verbose mode.\n");
	fprintf(stderr,"       -s dump statistics before terminating.\n");
	fprintf(stderr,"  Environment:\n");
	fprintf(stderr,"       FCOM_MC_PREFIX defines multicast prefix (overridden by -p)\n");
	fprintf(stderr,"       FCOM_MC_IFADDR defines address of IF to be listened on\n");
}

int
main(int argc, char **argv)
{
int             ch;
int             rval    = 1;
unsigned        tout_ms = 1000;
int             async   = 0;
uint64_t        llid;
int             st;
char            *prefix = 0;
char            *mcifad = 0;
struct timespec tout;
FcomBlobRef     blob;
int             level   = 0;
uint32_t        ifaddr;
FcomBlobSetRef  set     = 0;
FcomID          *idnts  = 0;
int             nids    = 0;
int             nsubs   = 0;
int             stats   = 0;
int             bufs    = 10;
int             i;

	while ( (ch = getopt(argc, argv, "ab:hp:st:v")) >= 0 ) {
		switch (ch) {
			case 'h':
				rval = 0;
			default:
				usage(argv[0]);
				return rval;

			case 'a':
				async = 1;
			break;

			case 'b':
				if ( 1 != sscanf(optarg, "%i", &bufs) || bufs < 1) {
					fprintf(stderr,"Invalid arg to -b: must be positive # of buffers\n");
					usage(argv[0]);
					return 1;
				}
			break;

			case 's':
				stats = 1;
			break;

			case 't':
				if ( 1 != sscanf(optarg,"%u",&tout_ms) ) {
					fprintf(stderr,"Invalid 'timout_ms' argument\n");
					usage(argv[0]);
					return 1;
				}
			break;

			case 'p':
				prefix = optarg;
			break;

			case 'i':
				mcifad = optarg;
			break;

			case 'v':
				level = 1;
			break;
		}
	}

	if ( !prefix && !(prefix = getenv("FCOM_MC_PREFIX")) ) {
		fprintf(stderr,"Missing FCOM multicast prefix. Use '-p' option of define FCOM_MC_PREFIX env-var\n");
		return 1;
	}

	if ( !mcifad ) {
		mcifad = getenv("FCOM_MC_IFADDR");
		/* Don't complain if NULL - OK if routing tables are used
		 * or only a single NIC present.
		 */
	}

	nids = argc - optind;

	if ( nids <= 0 ) {
		fprintf(stderr,"Missing or non-numeric FCOM blob ID\n");
		usage(argv[0]);
		return 1;
	}

	if ( ! (idnts = malloc( nids * sizeof(FcomID) )) ) {
		fprintf(stderr,"Not enough memory (for ID array)\n");
		return 1;
	}

	for ( i=0; i<nids; i++ ) {
 		if (  1 != sscanf(argv[optind+i],"%"SCNi64,&llid) ) {
			fprintf(stderr,"Non-numeric FCOM blob ID (#%i)\n", i+1);
			usage(argv[0]);
			goto bail;
		}
		idnts[i] = (FcomID)llid;
	}

	if ( mcifad ) {
		if ( INADDR_NONE == (ifaddr = inet_addr(mcifad)) ) {
			fprintf(stderr,"Invalid IP address: %s\n", mcifad);
			goto bail;
		}
		udpCommSetIfMcastInp(ifaddr);
	}

	fcom_silent_mode = 1;
	st = fcomInit( prefix, bufs );

	if ( st ) {
		fprintf(stderr,"Unable to initialize FCOM: %s\n", fcomStrerror(st));
		goto bail;
	}

	if ( nids > 1 )
		async = 1;

	for ( i=0; i<nids; i++ ) {
		st = fcomSubscribe( idnts[i], async ? FCOM_ASYNC_GET : FCOM_SYNC_GET );

		if ( FCOM_ERR_UNSUPP == st ) {
			fprintf(stderr,"Warning: synchronous get not supported; using asynchronous mode\n");
			async = 1;
			st    = fcomSubscribe( idnts[i], FCOM_ASYNC_GET );
		}

		if ( st ) {
			fprintf(stderr,"FCOM subscription failed: %s\n", fcomStrerror(st));
			goto bail;
		}
		nsubs++;
	}

	if ( nids > 1 ) {
		/* build a set */
		st = fcomAllocBlobSet( idnts, nids, &set );
		if ( st ) {
			fprintf(stderr,"fcomAllocBlobSet failed: %s\n", fcomStrerror(st));
			set = 0;
			goto bail;
		}

		st = fcomGetBlobSet( set, 0, (1<<nids) - 1, FCOM_SET_WAIT_ALL, tout_ms );
		if ( st ) {
			fprintf(stderr,"fcomGetBlobSet failed: %s\n", fcomStrerror(st));
			if ( FCOM_ERR_TIMEDOUT != st )
				goto bail;
		}

		for ( i=0; i<nids; i++ ) {
			fcomDumpBlob( set->memb[i].blob, level, stdout );
		}
	} else {
		if ( async ) {
			tout.tv_sec  = tout_ms/1000;
			tout.tv_nsec = (tout_ms - tout.tv_sec*1000)*1000000;
			if ( nanosleep( &tout, 0 ) ) {
				fprintf(stderr,"Sleep aborted\n");
				goto bail;
			}
		}

		st = fcomGetBlob( idnts[0], &blob, async ? 0 : tout_ms );
		if ( st ) {
			fprintf(stderr,"fcomGetBlob(%sSYNCH) failed: %s\n", async ? "A":"", fcomStrerror(st));
			goto bail;
		}
		st = fcomDumpBlob( blob, level, stdout );
		fcomReleaseBlob(&blob);
		if ( st < 0 ) {
			fprintf(stderr,"fcomDumpBlob failed: %s\n", fcomStrerror(st));
			goto bail;
		}
	}

	rval = 0;

bail:

	if ( stats )
		fcomDumpStats(stdout);

	if ( set ) {
		st = fcomFreeBlobSet( set );
		if ( st )
			fprintf(stderr,"fcomFreeBlobSet failed: %s\n", fcomStrerror(st));
	}

	for ( i=0; i<nsubs; i++ ) {
		st = fcomUnsubscribe( idnts[i] );
		if ( st )
			fprintf(stderr,"fcomUnsubscribe("FCOM_ID_FMT") failed: %s\n", idnts[i], fcomStrerror(st));
	}

	free( idnts );

	return rval;
}
